### PR TITLE
ConstantTimeEq proc macro derive

### LINKED ISSUE #280: Impl ConstantTimeEq on quote types
Comment from https://github.com/mobilecoinfoundation/attestation/pull/3#discussion_r1116260392
> I'd say we should just impl ConstantTimeEq for Attributes in core-types, honestly.

### PR DESCRIPTION
**Pull request by varsha888**
### Motivation
Fixes #280 
Using [Subtle crate](https://crates.io/crates/subtle)
Implemented ConstantTimeEq using proc macro derive 
Implemented ConstantTimeGreater and ConstantTimeLesser

### Future Work
Move proc macro into subtle crate

### CODE REVIEW

File: core/types/Cargo.toml [hunk 1]
```diff
@@ -18,12 +18,14 @@ alloc = []
...
 [dependencies]
 bitflags = "2.0.0"
+constant_time_derive = { path = "../../constant_time_derive", version = "0.1.0" }
 displaydoc = { version = "0.2.3", default-features = false }
 mc-sgx-core-sys-types = { path = "../sys/types", version = "=0.5.1-beta.0" }
 mc-sgx-util = { path = "../../util", version = "=0.5.1-beta.0" }
 nom = { version = "7.1.2", default-features = false }
 rand_core = { version = "0.6.4", default-features = false }
 serde = { version = "1.0.152", default-features = false, features = ["derive"], optional = true }
+subtle = { version = "2.4.0", default-features = false, features = ["i128"] }
```
Comment (nick-mobilecoin) [inline core/types/Cargo.toml:28]: :question: I'm not sure we want to pull in i128, it's not available on all platforms and we want the types to work on HW wallets.
It may be that we want CT behind feature flags, but it still seems like it might be an attack vector if a malicious app times how long it takes a HW wallet to verify.

File: constant_time_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+fn derive_ct_eq(input: &DeriveInput) -> TokenStream {
+    let ident = &input.ident;
+    let data = &input.data;
+    let generics = &input.generics;
+
+    let is_lifetime = parse_lifetime(generics);
+    let ct_eq_stream: proc_macro2::TokenStream =
+        parse_data(data).expect("Failed to parse DeriveInput data");
+    let data_ident = if is_lifetime {
```
Comment (NotGyro) [inline constant_time_derive/src/lib.rs:82]: Does this break on types with multiple explicit lifetimes?

File: core/build/src/lib.rs [hunk 3]
```diff
@@ -176,6 +178,20 @@ impl SgxParseCallbacks {
         self
     }
 
+    /// Types to derive constant time for
+    ///
+    /// # Arguments
+    /// * `default_types` - Types to derive default for.
```
Comment (nick-mobilecoin) [inline core/build/src/lib.rs:184]: :wrench: Wrong docs, though one could probably omit the arguments from the docs as it's named clearly and I think the other comments were me just being pedantic. It looks like rust documentation suggest not documenting arguments if they are clearly communicated by name.
```suggestion
    /// * `constant_time_types` - Types to derive constant time for.
```

File: core/types/src/attestation_key.rs [hunk 4]
```diff
@@ -251,6 +253,91 @@ mod test {
...
+    }
+
+    #[test]
+    fn ct_eq_version() {
+        let first = Version(2);
+        let second = Version(2);
+        let choice_result = first.ct_eq(&second);
+        let result: bool = From::from(choice_result);
+        assert!(result);
+    }
```
Comment (nick-mobilecoin) [inline core/types/src/attestation_key.rs:272]: :thought_balloon: I'm not sure tests at this level are very beneficial:
1. These are only testing `==` nothing is testing for `!=`.
2. They feel pretty noisy here and don't provide confidence on the full logic of the macro itself

Suggest making a `tests` directory in the macro crate and test the different types there.
This will:
- make it easier to test boundary conditions there
- make the tests live with the crate if it gets moved somewhere else
- prevents noise here. I know that we're testing the to/from logic. but we're not testing hash, ord, etc here.
```rust
use subtle::ConstantTimeEq;
use constant_time_derive::ConstantTimeEq;

#[derive(ConstantTimeEq)]
struct NamedFields {
    a: u8,
    b: u8,
}

#[test]
fn named_fields() {
    let a = NamedFields { a: 1, b: 2 };
    let b = NamedFields { a: 1, b: 2 };
    let c = NamedFields { a: 1, b: 3 };
    assert!(bool::from(a.ct_eq(&b)));
    assert!(bool::from(!a.ct_eq(&c)));
}

```

File: constant_time_derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    }
+                }
+            }
+            Ok(token_stream)
+        }
+        Fields::Unnamed(_) => {
+            let token_stream = quote! { {self.0}.ct_eq({&other.0}) };
+
+            Ok(token_stream)
+        }
```
Comment (nick-mobilecoin) [inline constant_time_derive/src/lib.rs:34]: :wrench: TBH I'm not sure exactly what's going on here, but i think it only looks at the first field
```rust
#[derive(ConstantTimeEq)]
struct UnnamedFields(u8, u8);

#[test]
fn unnamed_fields() {
    let a = UnnamedFields(3, 4);
    let b = UnnamedFields(3, 4);
    let c = UnnamedFields(3, 3);
    assert!(bool::from(a.ct_eq(&b)));
    assert!(bool::from(!a.ct_eq(&c)));
}
```
That test fails with
```console
assertion failed: bool::from(!a.ct_eq(&c))
thread 'unnamed_fields' panicked at 'assertion failed: bool::from(!a.ct_eq(&c))', constant_time_derive/tests/newtype_struct.rs:29:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: newtype_struct::unnamed_fields
             at ./tests/newtype_struct.rs:29:5
   4: newtype_struct::unnamed_fields::{{closure}}
             at ./tests/newtype_struct.rs:24:21
   5: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Review by nick-mobilecoin — COMMENTED** [review]
submitting without specific approval/denial since I'll be awol for the next couple of weeks

**Review by jcape — CHANGES_REQUESTED** [review]

File: constant_time_derive/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "constant_time_derive"
+version = "0.1.0"
+edition = "2021"
+license = "Apache-2.0"
```
[the remark this answers is not in the document]
Comment (varsha888) [reply constant_time_derive/Cargo.toml:5]: It should be ok as proc macro for constantTimeEq will move into subtle.

File: core/types/Cargo.toml (hunk 1 above)
Comment (varsha888) [reply core/types/Cargo.toml:28]: @jcape said we can change it if we come across a problem. @jcape thoughts?

File: core/build/src/lib.rs (hunk 3 above)
Comment (varsha888) [reply core/build/src/lib.rs]: Good catch. Changed it

File: constant_time_derive/src/lib.rs (hunk 5 above)
Comment (varsha888) [reply constant_time_derive/src/lib.rs]: I misunderstood that unnamed structs will be for the FFI types. Code should handle this now.

File: constant_time_derive/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            panic!("Cannot derive ct_eq for fields in enums")
+        }
+    }
+    let token_stream = quote! {
+        if self == other {
+            ::subtle::Choice::from(1)
+        }
+        else {
+            ::subtle::Choice::from(0)
+        }
```
[the remark this answers is not in the document]
Comment (varsha888) [reply constant_time_derive/src/lib.rs:64]: Choice doesnt not have from impl for bool. It has to be added into subtle to do this. I can include it in my PR to subtle.

File: constant_time_derive/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            while let Some(field) = iter.next() {
+                let ident = &field.ident;
+                match iter.peek() {
+                    None => token_stream.extend(quote! { {self.#ident}.ct_eq(&{other.#ident}) }),
+                    Some(_) => {
+                        token_stream.extend(quote! { {self.#ident}.ct_eq(&{other.#ident}) & })
+                    }
+                }
+            }
+            Ok(token_stream)
```
Comment (nick-mobilecoin) [inline constant_time_derive/src/lib.rs:29]: :clown_face: I don't know that this suggestion is much better. The while loop and peeking for "is there more" felt off
```suggestion
            let mut token_stream = quote!();
            let separators = ::core::iter::once("").chain(::core::iter::repeat("&"));
            let items = separators.zip(fields_named.named.iter());
            for (separator, field) in items {
                let ident = &field.ident;
                let separator_token: proc_macro2::TokenStream = separator.parse().expect("Should be a valid token");
                token_stream.extend(quote! { #separator_token {self.#ident}.ct_eq(&{other.#ident}) });
            }
            Ok(token_stream)
```

File: constant_time_derive/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        for _ in 1..is_lifetime {
+            s.push_str(", '_");
+        }
+        s.push('>');
+
+        s
+    } else {
+        ident.to_string()
+    };
+    let ident_stream: proc_macro2::TokenStream = data_ident.parse().unwrap();
```
Comment (nick-mobilecoin) [inline constant_time_derive/src/lib.rs:107]: :upside_down_face: We've been trying to use expects over unwraps as the "reason" message will often play the same role as a comment, it also makes it easier to automatically check for no use of unwrap
```suggestion
    let ident_stream: proc_macro2::TokenStream =
        data_ident.parse().expect("Should be valid lifetime tokens");
```

File: core/types/src/quote.rs [hunk 10]
```diff
@@ -394,4 +396,208 @@ mod test {
...
+
+    #[test]
+    fn ct_eq_quoting_enclave_report_info() {
+        let nonce = sgx_quote_nonce_t { rand: [1u8; 16] };
+        let app_enclave_target_info = sgx_target_info_t {
+            mr_enclave: Default::default(),
+            attributes: Default::default(),
+            reserved1: [1u8; 2],
+            config_svn: 0,
+            misc_select: 0,
```
Comment (nick-mobilecoin) [inline core/types/src/quote.rs:408]: :question: I think we can come back later and decide, but I wonder if the reserved fields should be excluded

File: core/types/src/svn.rs [hunk 11]
```diff
@@ -2,41 +2,73 @@
...
 
 impl_newtype_for_bytestruct! {
     CpuSvn, sgx_cpu_svn_t, SGX_CPUSVN_SIZE, svn;
 }
 
+impl ConstantTimeGreater for CpuSvn {
+    fn ct_gt(&self, other: &Self) -> Choice {
+        if !self.0.svn.len().eq(&other.0.svn.len()) {
+            return Choice::from(0);
+        }
```
Comment (nick-mobilecoin) [inline core/types/src/svn.rs:56]: :wrench: This check isn't needed since the array is defined with a size inside of a struct
```suggestion
```

**Review by nick-mobilecoin — CHANGES_REQUESTED** [review]
The main thing that should be done is mapping the `==` to Choice in the enum derive logic.
the other stuff is mainly nits and suggestions

File: constant_time_derive/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        if let Fields::Unnamed(_) = variant.fields {
+            panic!("Cannot derive ct_eq for fields in enums")
+        }
+    }
+    let token_stream = quote! {
+        if self == other {
+            ::subtle::Choice::from(1)
+        }
+        else {
+            ::subtle::Choice::from(0)
```
[the remark this answers is not in the document]
Comment (varsha888) [reply constant_time_derive/src/lib.rs:63]: Interesting, Java or Kotlin didnt do this.

File: constant_time_derive/src/lib.rs (hunk 9 above)
Comment (varsha888) [reply constant_time_derive/src/lib.rs:107]: 👍

File: core/types/src/quote.rs (hunk 10 above)
Comment (varsha888) [reply core/types/src/quote.rs:408]: Why should it be excluded?

File: constant_time_derive/src/lib.rs (hunk 8 above)
Comment (varsha888) [reply constant_time_derive/src/lib.rs:29]: I felt peeking is more readable but thanks for this, I might use this elsewhere

File: core/types/src/quote.rs (hunk 10 above)
Comment (nick-mobilecoin) [reply core/types/src/quote.rs:408]: Often reserved fields are for future expansion(and memory alignment). Often times they're whatever happened to be on the stack for the structure.

**Review by nick-mobilecoin — APPROVED** [review]

**Comment by nick-mobilecoin** [comment]
closing due to shifting focus
